src/hb-aat-layout-lcar-table.hh

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright © 2018  Ebrahim Byagowi
+ *
+ *  This is part of HarfBuzz, a text shaping library.
+ *
+ * Permission is hereby granted, without written agreement and without
+ * license or royalty fees, to use, copy, modify, and distribute this
+ * software and its documentation for any purpose, provided that the
+ * above copyright notice and the following two paragraphs appear in
+ * all copies of this software.
+ *
+ * IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE TO ANY PARTY FOR
+ * DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
+ * ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN
+ * IF THE COPYRIGHT HOLDER HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * THE COPYRIGHT HOLDER SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING,
+ * BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
+ * ON AN "AS IS" BASIS, AND THE COPYRIGHT HOLDER HAS NO OBLIGATION TO
+ * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+ */
+#ifndef HB_AAT_LAYOUT_LCAR_TABLE_HH
+#define HB_AAT_LAYOUT_LCAR_TABLE_HH
+
+#include "hb-open-type.hh"
+#include "hb-aat-layout-common.hh"
+
+/*
+ * lcar -- Ligature caret
+ * https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6lcar.html
+ */
+#define HB_AAT_TAG_lcar HB_TAG('l','c','a','r')
+
+
+namespace AAT {
+
+struct lcar
+{
+  static const hb_tag_t tableTag = HB_AAT_TAG_lcar;
+
+  inline unsigned int get_lig_carets (hb_font_t      *font,
+				      hb_direction_t  direction,
+				      hb_codepoint_t  glyph,
+				      unsigned int    start_offset,
+				      unsigned int   *caret_count /* IN/OUT */,
+				      hb_position_t  *caret_array /* OUT */) const
+  {
+    const ArrayOf<HBINT16>& array = this+(*lookup.get_value (glyph, font->face->num_glyphs));
```

Comment:
`get_value()` can return nullptr.  Use `get_value_or_null()`.

Reply:
Typically when there's a pointer, it means it can be null.  When there's a reference, not.

Reply:
Thanks.

`this+lookup.get_value_or_null (glyph, font->face->num_glyphs);` fails with

```
hb-aat-layout-common.hh: In instantiation of ‘const typename T::type AAT::LookupFormat10<T>::get_value_or_null(hb_codepoint_t) const [with T = OT::OffsetTo<OT::ArrayOf<OT::IntType<short int, 2> > >; typename T::type =OT::IntType<short unsigned int, 2>; hb_codepoint_t = unsigned int]’:
hb-aat-layout-common.hh:353:61:   required from ‘const typename T::type AAT::Lookup<T>::get_value_or_null(hb_codepoint_t, unsigned int) const [with T = OT::OffsetTo<OT::ArrayOf<OT::IntType<short int, 2> > >; typename T::type = OT::IntType<short unsigned int, 2>; hb_codepoint_t = unsigned int]’
hb-aat-layout-lcar-table.hh:50:97:   required from here
hb-aat-layout-common.hh:310:12: error: could not convert ‘v’ from ‘unsigned int’ to ‘const type {aka const OT::IntType<short unsigned int, 2>}’
     return v;
            ^
```

Manually doing the same.

Reply:
Oh right.  Something to eventually fix, such that format10 can work as well.  But adding cast constructors to Offset type currently fails on pre-C++11 I think.  Maybe in a separate PR add an explicit operator cast to numeric types and see if it passes bots.

Reply:
https://github.com/harfbuzz/harfbuzz/pull/1370